Add a parser for NIDS alert data in the fast-format

jgru opened the pull request:
## One line description of pull request
Add a parser for NIDS alert data in the so-called fast-format, which is used by Snort3 as well as Suricata.

## Description:

### Background
This PR adds a parser for NIDS alert log data in the so-called *fast* format which is supported by Snort3 and Suricata.
Alerts are logged in a single line format, as shown below: 

`(Year/)?Month/Day-Hour:Minute:Second:FractionOfSeconds [**] [int:int:int] string [**]  [Classification: string] [Priority : int] {protocol} SRC_IP:SRC_PORT -> DST_IP:DST_PORT` 

Note that Snort3 and Suricata use slightly different "dialects", e.g., Suricata logs the year, which Snort3 does not. 

Those log files are commonly stored at the following locations:
- `/var/log/snort/alert_fast.txt`
- `/var/log/suricata/fast.log`

The incorporation of those logs in Plaso might be helpful in two ways:
First, it might support investigations of Linux servers running one of the before-mentioned NIDS, and second, it might allow correlating intrusion activity on endpoints with alert log data captured. 

### Technicalities
The following files have been created or modified:

- `plaso/parsers/nids_fast.py`
- `tests/parsers/nids_fast.py`
- `test_data/snort3_alert_fast.log`
- `test_data/suricata_alert_fast.log`
- `data/formatters/generic.yaml`

All the parsing is realized in the module `plaso/parsers/nids_fast.py`. To parse fast log data, `text_parser.PyparsingSingleLineTextParser` has been subclassed, and the necessary grammar has been defined with the help of `pyparsing` and pre-defined parsing rules in `text_parser`.  As a result, events of type `nids:alert:fast` (implemented in the class `NIDSFastAlertEventData`) are created. 

Beside the unit tests in `tests/parsers/nids_fast.py`, the intended functionality has been verified by running the following command in a newly created virtual environment: 
```
venv/bin/log2timeline.py --storage_file test.plaso test_data/snort3_alert_fast.log
venv/bin/psort.py --output_time_zone "UTC" -o "dynamic" -w test.csv test.plaso
```

**Related issue (if applicable):** fixes #4219

## Checklist:
  - [x] Automated checks (Travis, Codecov, Codefactor )pass
  - [x] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
  - [x] Reviewer assigned

joachimmetz commented on plaso/parsers/nids_fast.py:
```diff
@@ -0,0 +1,304 @@
+class NIDSFastParser(tp.PyparsingSingleLineTextParser):
+    """NIDS fast.log parser."""
+
+    NAME = "fast_alert_log"
```
Please take a look at https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html

joachimmetz replied:
Please use single quote for a single line string (and double for multi line strings)

jgru replied:
> Please take a look at https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html

Addressed in 557d23435b855fa7d1a59fab45124ebf837636b9.

> Please use single quote for a single line string (and double for multi line strings)

Addressed in 9a7198c323074813796d49cdb0365fde3438f628.

joachimmetz commented:
@jgru thx for the proposed changes, note that Plaso submissions follow a style guide. PTAL at https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html

joachimmetz commented on plaso/parsers/nids_fast.py:
```diff
@@ -0,0 +1,304 @@
+    DATA_FORMAT = "NIDS fast alert log"
+
+    _FASTLOG_VERIFICATION_PATTERN = (
+        r"^(\d{2}\/)?\d{2}\/\d{2}\-\d{2}:\d{2}:\d{2}.\d{6}\s*\[\*\*\]\s*\[\d*:\d"
```
Is there a way to structure this in a more maintainable way?

jgru replied:
I tried to improve the maintainability of this regex by splitting it into logical parts and adding comments in 4bdcaecf29bf3ec35ed876c3de35ad2bde6dab3b.

joachimmetz replied:
thx, yes this makes it much easier to read. I'll have some more thoughts on how to improve this in the future, but these changes are sufficient for now.

joachimmetz commented on plaso/parsers/nids_fast.py:
```diff
@@ -0,0 +1,304 @@
+
+
+class NIDSFastParser(tp.PyparsingSingleLineTextParser):
+    """NIDS fast.log parser."""
```
Plaso uses 2 space indentation

jgru replied:
Addressed in 66015c33f8078ac2351cb1706f3111ec86734de8.

jgru commented on plaso/parsers/nids_fast.py:
```diff
@@ -0,0 +1,304 @@
+    """Fast log event data.
+
+    Attributes:
+      sid (str): identifier of the rule that triggered the alert,
```
Adressed in c2c690453a770f11229775139757be15b1dbbb30 and 842be03c4fc917bb8c3c969cc64a26042f37e25c.

jgru commented:
> @jgru thx for the proposed changes, note that Plaso submissions follow a style guide. PTAL at https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html

Thank you very much for looking at my PR, @joachimmetz. I had another look at the style guide and addressed your comments. 
I hope that the issues with the coding style are resolved now.

joachimmetz commented on data/formatters/generic.yaml:
```diff
@@ -462,6 +462,24 @@ short_message:
+- 'Destination Port: {destination_port}'
+- 'Protocol: {protocol}'
+- 'Classification: {classifciation}'
+- 'Rule ID: {rule_id}'
```
nit: per style guide don't use abbreviations

jgru replied:
Changed in 654138576ceba8f8a7eff21dbbe4c8d2b9b31c5d.

joachimmetz commented:
@jgru having a closer look and I'll make some changes

joachimmetz commented on data/formatters/generic.yaml:
```diff
@@ -462,6 +462,24 @@ short_message:
+- 'Message: {message}'
+- 'Source-IP: {source_ip}'
+- 'Destination-IP: {destination_ip}'
+- 'Rule ID: {sid}'
```
I assume this should be rule_id

jgru replied:
Changed in 9412614b863062cb598cb06c6acbf84d04d82f71.

joachimmetz commented on data/formatters/generic.yaml:
```diff
@@ -462,6 +462,24 @@ short_message:
 - '{file_md5}'
 ---
 type: 'conditional'
+data_type: 'nids:alert:fast'
```
I'll rename this to snort, to disambiguate with other NIDS

jgru replied:
Okay, the parser can process Suricata's `fast.log` as well, but I'm certainly fine with it since it all goes back to snort. 
Changed in 0b5ab6edd7244e8315933ec6a828679672ac7f38. 

Do you think the module and class names should be renamed as well?

joachimmetz replied:
IMHO yes, but changes are on the way to also make this a text (parser) plugin

joachimmetz commented on plaso/parsers/nids_fast.py:
```diff
@@ -0,0 +1,327 @@
+from plaso.containers import events, time_events
+from plaso.lib import definitions, errors
+from plaso.parsers import manager
+from plaso.parsers import text_parser as tp
```
nit: don't use abbreviations

jgru replied:
Changed in dbcd264507074c60af2de80c85be7edf29151ab5.

joachimmetz commented on plaso/parsers/nids_fast.py:
```diff
@@ -0,0 +1,327 @@
+import pyparsing
+from dfdatetime import time_elements as dfdatetime_time_elements
+from plaso.containers import events, time_events
+from plaso.lib import definitions, errors
```
nit: single import per line

jgru replied:
Changed in 5759084c0311433ffd012dc583dadece99368f3f.

joachimmetz commented on plaso/parsers/nids_fast.py:
```diff
@@ -0,0 +1,327 @@
+  """Fast log event data.
+
+  Attributes:
+    sid (str): identifier of the rule that triggered the alert,
```
nit: sid => rule_id

jgru replied:
Changed in 399deea69c234a3de55893ff06ce14f26c253bb5.

joachimmetz commented on plaso/parsers/nids_fast.py:
```diff
@@ -0,0 +1,327 @@
+      # Optional destination port
+      r'(:\d*)?$'
+  )
+  _VERIFICATION_REGEX = re.compile(_FASTLOG_VERIFICATION_PATTERN)
```
This can be combined with the _FASTLOG_VERIFICATION_PATTERN definition.

jgru replied:
Changed in 8d6e6f7f7b1a623455c5d0c4cb2359f04bbd6748.

joachimmetz commented on plaso/parsers/nids_fast.py:
```diff
@@ -0,0 +1,327 @@
+  )
+  _VERIFICATION_REGEX = re.compile(_FASTLOG_VERIFICATION_PATTERN)
+
+  _PYPARSING_COMPONENTS = {
```
The double indirection is not needed IMHO

jgru replied:
Could you explain what you want to be changed here exactly?

joachimmetz replied:
no need to make changes, I'm making some also for the recent text plugin change
but what I mean there is no need to use a dict like _PYPARSING_COMPONENTS and then doing another lookup to get the pyparsing definition, I opt to use static class definitions.

joachimmetz commented on plaso/parsers/nids_fast.py:
```diff
@@ -0,0 +1,327 @@
+              pyparsing.Word(pyparsing.printables, exclude_chars='["')
+              | pyparsing.White(" ", max=2)
+          )
+      ).set_results_name('message'),
```
I assume you mean setResultsName ?

jgru replied:
Yes, changed in fb491fa71f7db0725a9c96c0c24f8fd0805f64e0.

joachimmetz commented on plaso/parsers/text_plugins/snort_fastlog.py:
```diff
@@ -0,0 +1,300 @@
+    """Initializes a parser."""
+    super(SnortFastLogTextPlugin, self).__init__()
+    self._last_month = 0
+    self._maximum_year = 0
```
note to self _maximum_year and _plugin_by_reporter are not used, and to reset _year_use and _last_month on verification

joachimmetz commented:
> Thank you very much for looking at my PR, @joachimmetz. I had another look at the style guide and addressed your comments.
> I hope that the issues with the coding style are resolved now.

Noprob, takes a couple of iterations to get accustomed to the style.

Have a look, I've made some changes, also seeing recent text parser plugin changes in the codebase.

jgru commented:
> Noprob, takes a couple of iterations to get accustomed to the style.
> 

Indeed, it's a lot different from current mainstream style.  

> Have a look, I've made some changes, also seeing recent text parser plugin changes in the codebase.

Thank you very much for your help on this. It looks way better now.

joachimmetz commented on plaso/parsers/text_plugins/snort_fastlog.py:
```diff
@@ -0,0 +1,303 @@
+      r'\-\>\s*'
+      # Destination IPv6-address
+      r'((([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:'
+      r'|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a -fA-F]{1,4}:){1,5}'
```
I assume this should be `0-9a-fA-F` not `0-9a -fA-F`

jgru replied:
Yes, absolutely. Thanks for correction in 7881955496450b9d78afbede9b2740247ec55766.

joachimmetz reviewed (approved):
LGTM
